js/store.js

```diff
@@ -112,51 +216,139 @@ const store = {
+    }
+    return false;
   },
 
   setFirstParty(hostname, data) {
     if (!hostname) {
       throw new Error('setFirstParty requires a valid hostname argument');
     }
+    const isNewWebsite = this.isNewWebsite(hostname);
+
     this.setWebsite(hostname, data);
+
+    if (isNewWebsite) {
+      this.updateChild(this.outputWebsite(hostname, this._websites[hostname]));
+    }
   },
 
+  // @todo use data from capture to update UI/filter graph
+  // eslint-disable-next-line no-unused-vars
   setThirdParty(origin, target, data) {
-    let newThirdParty = false;
+    let isNewWebsite = false;
     if (!origin) {
       throw new Error('setThirdParty requires a valid origin argument');
     }
 
-    const firstParty = this.getWebsite(origin);
-    const thirdParty = this.getWebsite(target);
+    let firstParty = this.getWebsite(origin);
+    let thirdParty = this.getWebsite(target);
 
     if (!('thirdPartyHostnames' in firstParty)) {
       firstParty['thirdPartyHostnames'] = [];
     }
     if (!firstParty['thirdPartyHostnames'].includes(target)) {
-      firstParty['thirdPartyHostnames'].push(target);
-      newThirdParty = true;
+      // third party is not currently linked to first party
+      ({ firstParty, thirdParty, isNewWebsite }
+        = this.whiteListCheck(origin, target, firstParty, thirdParty));
     }
-    if (!('firstPartyHostnames' in firstParty)) {
+    if (!('firstPartyHostnames' in thirdParty)) {
       thirdParty['firstPartyHostnames'] = [];
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
       thirdParty['firstPartyHostnames'].push(origin);
-      newThirdParty = true;
     }
 
-    this.setWebsite(origin, firstParty);
+    this.setFirstParty(origin, firstParty);
     this.setWebsite(target, thirdParty);
 
-    if (newThirdParty) {
-      this.updateChild(this.outputWebsite(target, data, origin));
+    if (isNewWebsite && thirdParty['isVisible']) {
+      this.updateChild(this.outputWebsite(target, this._websites[target]));
     }
   },
 
+  // Checks if third party is on whitelist for first party,
+  // and if the whitelisted third party should be unwhitelisted.
+  whiteListCheck(origin, target, firstParty, thirdParty) {
+    let isNewWebsite = false;
+    if (!('isVisible' in thirdParty)) {
+      if (this.onWhiteList(origin, target)) {
+        // third party is whitelisted for this first party
+        // ignore this third party for now.
+        thirdParty['isVisible'] = false;
+      } else {
+        thirdParty['isVisible'] = true;
+        firstParty['thirdPartyHostnames'].push(target);
+        isNewWebsite = true;
+      }
+    } else if (!(this.onWhiteList(origin, target))) {
+      if (!thirdParty['isVisible']) {
+        // third party was previously whitelisted, and its
+        // previously ignored link with the owning first party
+        // should be added now
+        thirdParty['isVisible'] = true;
+        const whiteListedFirstParty = thirdParty['firstPartyHostnames']
```

Comment:
@biancadanforth google analytics is whitelisted by google.de, google.co.uk abc.xyz what am I missing?